src/librustc_data_structures/temp_dir.rs

```diff
@@ -0,0 +1,34 @@
+use std::mem::ManuallyDrop;
+use std::path::Path;
+use tempfile::TempDir;
+
+/// This is used to avoid TempDir being dropped on error paths unintentionally.
+#[derive(Debug)]
+pub struct MaybeTempDir {
+    dir: ManuallyDrop<TempDir>,
+    // Whether the TempDir should be deleted on drop.
+    keep: bool,
+}
+
+impl Drop for MaybeTempDir {
+    fn drop(&mut self) {
+        // Safety: We are in the destructor, and no further access will
+        // occur.
+        let dir = unsafe { ManuallyDrop::take(&mut self.dir) };
```

Comment:
No need for `unsafe`. Just `drop(self.dir.into_inner())` if `keep` is false.

Reply:
Oh NVM, we can't move out of it in the destructor. I would probably have done this with `ManuallyDrop::drop`, but that's also unsafe.

Reply:
I considered that, but then we need to have two unsafe blocks I think - one for the into_path case, with take, the other with drop. Seems more complicated.

Reply:
...otherwise we leak the `PathBuf` inside `TempDir`. Whoops. I'm all caught up now.